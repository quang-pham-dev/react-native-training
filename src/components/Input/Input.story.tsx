import React from 'react'
import {storiesOf} from '@storybook/react-native'
import {action} from '@storybook/addon-actions'
import {StoryScreen, Story, UseCase} from '../../../storybook/views'
import {text} from '@storybook/addon-knobs'

import {Icons} from '@themes'
import Input, {InputType} from './index'

declare let module: NodeModule

storiesOf('Input', module)
  .addDecorator(fn => <StoryScreen>{fn()}</StoryScreen>)
  .add('Type normal', () => (
    <Story>
      <UseCase text={'normal'}>
        <Input
          label={text('label', 'Username')}
          placeholder={text('placeholder', 'Enter your username')}
          icon={Icons.filledProfile}
          value={''}
          type={InputType.NORMAL}
          onChange={action('onChange text')}
        />
      </UseCase>
    </Story>
  ))
  .add('Type password', () => (
    <Story>
      <UseCase text={'password'}>
        <Input
          label={text('label', 'Password')}
          placeholder={text('placeholder', 'Enter your password')}
          value={''}
          type={InputType.PASSWORD}
          onChange={action('onChange text')}
        />
      </UseCase>
    </Story>
  ))
