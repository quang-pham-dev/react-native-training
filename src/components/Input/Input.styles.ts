import styled from 'styled-components/native'

// Styles
import FlexStyled from '@components/Flex/Flex.styles'

// Themes
import {Fonts, Colors} from '@themes'

interface IconWrapperProps {
  mRight?: number
  mBottom?: number
}
interface TextInputProps {
  pLeft?: number
  pRight?: number
}

interface IInputStyled {
  Wrapper: typeof Wrapper
  LabelStyled: typeof LabelStyled
  InputField: typeof InputField
  IconEyeWrapper: typeof IconEyeWrapper
}

const Wrapper = styled(FlexStyled.RowCenterHorizontal)`
  border-bottom-width: 1px;
  border-bottom-color: ${Colors.palette.borderGray};
  width: 100%;

  ${props => props.pHorizontal && `padding-horizontal: ${props.pHorizontal}px`};
  ${props => props.pRight && `padding-right: ${props.pRight}px`};
`

const LabelStyled = styled.Text`
 font-size: ${Fonts.size.small}px;
 line-height:  ${Fonts.lineHeight.xs}px
 font-family: ${Fonts.type.interRegular};
 color:${Colors.palette.gray};
`

const InputField = styled.TextInput<TextInputProps>`
  flex: 1;
  padding-left: ${props => (props.pLeft ? props.pLeft : 0)}px;
  padding-right: ${props => (props.pRight ? props.pRight : 5)}px;
  padding-vertical: 15px;
  font-family: ${Fonts.type.interMedium};
  font-size: ${Fonts.size.normal}px;
  line-height: ${Fonts.lineHeight.md}px;
  color: ${Colors.palette.black};
`

const IconEyeWrapper = styled.TouchableOpacity<IconWrapperProps>`
  ${props => props.mRight && `margin-right: ${props.mRight}px;`}
  ${props => props.mBottom && `margin-bottom: ${props.mBottom}px;`}
`

const InputStyled: IInputStyled = {
  Wrapper,
  LabelStyled,
  InputField,
  IconEyeWrapper,
}

export default InputStyled
