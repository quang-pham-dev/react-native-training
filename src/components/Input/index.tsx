import React, {memo, useCallback, useState} from 'react'
import {
  NativeSyntheticEvent,
  Pressable,
  TextInputChangeEventData,
} from 'react-native'

// LIBS
import isEqual from 'react-fast-compare'
import Icon from 'react-native-vector-icons/Feather'

// Styles
import FlexStyled from '@components/Flex/Flex.styles'
import InputStyled from './Input.styles'

// Themes
import {Colors} from '@themes'

export enum InputType {
  NORMAL = 'normal',
  EMAIL = 'email',
  PHONE = 'phone',
  NUMBER = 'number',
  PASSWORD = 'password',
}

export enum KeyboardType {
  DEFAULT = 'default',
  NUMERIC = 'numeric',
  EMAIL_ADDRESS = 'email-address',
  ASCII_CAPABLE = 'ascii-capable',
  NUMBER_PUNCTUATION = 'numbers-and-punctuation',
  URL = 'url',
  NUMBER_PAD = 'number-pad',
  PHONE_PAD = 'phone-pad',
  NAME_PHONE_PAD = 'name-phone-pad',
  DECIMAL_PAD = 'decimal-pad',
  TWITTER = 'twitter',
  WEB_SEARCH = 'web-search',
  VISIBLE_PASSWORD = 'visible-password',
}

type InputProps = {
  type?: InputType
  value: string
  label?: string
  placeholder?: string
  icon?: React.ReactNode
  keyboardType?: KeyboardType
  autoFocus?: boolean
  onBlur?: () => void
  onChange: (value: string) => void | undefined
  onEndEditing?: (
    event: NativeSyntheticEvent<TextInputChangeEventData>,
    type: InputType,
  ) => void
  onClearInput?: () => void
  onSubmitEditing?: () => void
  inputProps?: {
    [key: string]: string | number
  }
}

const Input = ({
  type = InputType.NORMAL,
  value,
  label,
  placeholder,
  icon,
  keyboardType = KeyboardType.DEFAULT,
  onChange,
  onBlur,
  onSubmitEditing,
  inputProps,
}: InputProps) => {
  const isPassword = type === InputType.PASSWORD
  const [displayValue, setDisplayValue] = useState(value)
  const [showPassword, setShowPassword] = useState<boolean>(false)

  const handleTextChange = useCallback(
    (text: string) => {
      setDisplayValue(text)
      onChange(text)
    },
    [onChange, setDisplayValue],
  )

  const handleTogglePassword = (): void => setShowPassword(!showPassword)

  return (
    <FlexStyled.Default>
      {label && <InputStyled.LabelStyled>{label}</InputStyled.LabelStyled>}

      <InputStyled.Wrapper>
        <InputStyled.InputField
          type={type}
          value={displayValue}
          onChangeText={handleTextChange}
          onBlur={onBlur}
          autoCapitalize="none"
          underlineColorAndroid="transparent"
          onSubmitEditing={onSubmitEditing}
          placeholder={placeholder}
          keyboardType={keyboardType}
          secureTextEntry={isPassword && !showPassword}
          {...inputProps}
        />
        {isPassword && (
          <InputStyled.IconEyeWrapper onPress={handleTogglePassword}>
            {showPassword ? (
              <Icon name="eye-off" size={20} color={Colors.palette.gray} />
            ) : (
              <Icon name="eye" size={20} color={Colors.palette.gray} />
            )}
          </InputStyled.IconEyeWrapper>
        )}
        {icon && <Pressable>{icon}</Pressable>}
      </InputStyled.Wrapper>
    </FlexStyled.Default>
  )
}

export default memo(Input, isEqual)
