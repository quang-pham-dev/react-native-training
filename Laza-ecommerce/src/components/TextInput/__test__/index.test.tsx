import React from 'react';
import renderer from 'react-test-renderer';
import Input from 'components/TextInput';

describe('TextInput Component', () => {
  const onChangeText = jest.fn();
  const component = renderer.create(<Input value={''} onChangeText={onChangeText} />);
  test('should render correctly', () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
