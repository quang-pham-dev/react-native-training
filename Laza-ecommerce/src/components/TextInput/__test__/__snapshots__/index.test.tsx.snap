// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput Component should render correctly 1`] = `
<View
  style={
    Array [
      Object {
        "flexDirection": "row",
        "width": "100%",
      },
    ]
  }
>
  <TextInput
    allowFontScaling={true}
    autoCorrect={false}
    onChangeText={[MockFunction]}
    rejectResponderTermination={true}
    style={
      Array [
        Object {
          "borderBottomColor": "#E7E8EA",
          "borderBottomWidth": 1,
          "color": "#1D1E20",
          "flex": 1,
          "lineHeight": 16,
          "paddingVertical": 15,
        },
        Object {},
      ]
    }
    underlineColorAndroid="transparent"
    value=""
  />
</View>
`;
